classes/task/failed_task.php

```diff
@@ -31,6 +31,98 @@ public function get_name() {
+        $sort = '';
+        $fields = '*';
+        $limitfrom = 0;
+        $extractedevents = $DB->get_records('logstore_xapi_failed_log', $conditions, $sort, $fields, $limitfrom, $limitnum);
+        return $extractedevents;
+    }
+
+    private function extract_events($limitnum) {
+        global $DB;
+        $conditions = null;
+        $sort = '';
+        $fields = '*';
+        $limitfrom = 0;
+        $extractedevents = $DB->get_records('logstore_xapi_log', $conditions, $sort, $fields, $limitfrom, $limitnum);
+        return $extractedevents;
+    }
+
+    private function insert_failed_events_into_xapi_log($events) {
+        global $DB;
+        $DB->insert_records("logstore_xapi_log", $events);
+    }
+
+    private function delete_failed_events($events) {
+        global $DB;
+        $eventids = $this->get_delete_event_ids($events);
+        $DB->delete_records_list('logstore_xapi_failed_log', 'id', $eventids);
+    }
+
+    private function get_failed_events($events) {
+        $nonloadedevents = array_filter($events, function ($loadedevent) {
+            return $loadedevent['loaded'] === false;
+        });
+        $failedevents = array_map(function ($nonloadedevent) {
+            return $nonloadedevent['event'];
+        }, $nonloadedevents);
+
+        // enrich failed event with errortype and response
+        foreach ($failedevents as $event) {
+            if (!empty($_SESSION['xapi_errortype'])) {
+                $event->errortype = $_SESSION['xapi_errortype'];
+            }
+            if (!empty($_SESSION['xapi_response'])) {
+                $event->response = $_SESSION['xapi_response'];
+            }
+        }
+        return $failedevents;
+    }
+
+    private function get_successful_events($events) {
+        $loadedevents = array_filter($events, function ($loadedevent) {
+            return $loadedevent['loaded'] === true;
+        });
+        $successfulevents = array_map(function ($loadedevent) {
+            return $loadedevent['event'];
+        }, $loadedevents);
+        return $successfulevents;
+    }
+
+    private function store_failed_events($events) {
+        global $DB;
+        $failedevents = $this->get_failed_events($events);
+        $DB->insert_records('logstore_xapi_failed_log', $failedevents);
+        mtrace(count($failedevents) . " " . get_string('failed_events', 'logstore_xapi'));
+    }
+
+    private function record_successful_events($events) {
+        mtrace(count($this->get_successful_events($events)) . " " . get_string('successful_events', 'logstore_xapi'));
+    }
+
+    private function get_event_ids($loadedevents) {
+        return array_map(function ($loadedevent) {
+            return $loadedevent['event']->id;
+        }, $loadedevents);
+    }
+
+    private function get_delete_event_ids($loadedevents) {
+        $arr = array();
+        foreach ($loadedevents as $event) {
+            $arr[] = $event->id;
+        }
+        return $arr;
+    }
```

Comment:
what is the need for the `get_delete_event_ids` function @mlynn-lp, doesn't the above `get_event_ids` do the same?

Reply:
array_map throws a warning when the closure raises an exception

Reply:
In that case, I'm not sure we need both functions.
Happy to leave for now